### PR TITLE
Refactor(runtime/api_web): parse raw log events into structured ones 

### PR DESCRIPTION
**Pull request by EmileRolley**
Global refactoring of the log events by introducing _structured_ events and a parser to build them from _raw_ events -- previous `Runtime.event` type.

## Changelog

* New structured log event type:
  * The previous `event` type is now the `raw_event`.
  * New parser: `val Runtime.EventParser.parse_raw_events : raw_event list -> event list`.
  * New pretty printer: `pp_events`.
* Refactored web API with new functions: `retrieveEvents` and `retrieveRawEvents`.
* Input and output of function call now have types in logs (see [d82ecb5](https://github.com/CatalaLang/catala/pull/280/commits/d82ecb5f123a6d1e27b6a82350daba38d72c3b8a))
* Removes the `-s` flag from the `law_source/unit_tests/tests_allocations_familiales.ml` Makefile rule.

### CODE REVIEW

File: compiler/runtime.ml [hunk 1]
```diff
@@ -172,19 +172,46 @@ let embed_date x = Date x
...
+type raw_event =
+  | BeginCall of information
+  | EndCall of information
+  | VariableDefinition of information * runtime_value
   | DecisionTaken of source_position
 
-let log_ref : event list ref = ref []
+type event =
+  | VarDef of var_def
+  | VarDefWithFunCalls of var_def_with_fun_calls
```
Comment (denismerigoux) [inline compiler/runtime.ml:185]: I would be in favor of merging `VarDef` and `VarDefWithFunCall` into a single case `VarComputation`.

File: compiler/runtime.mli [hunk 2]
```diff
@@ -72,19 +72,137 @@ val embed_array : ('a -> runtime_value) -> 'a Array.t -> runtime_value
 
 (** {1 Logging} *)
 
+(** {2 Global process} *)
+
+(** The logging is constituted of two phases:
```
Comment (denismerigoux) [inline compiler/runtime.mli:77]: Amazing doc!

File: compiler/runtime.mli [hunk 3]
```diff
@@ -72,19 +72,137 @@ val embed_array : ('a -> runtime_value) -> 'a Array.t -> runtime_value
...
+  body : event list;
+  output : var_def_with_fun_calls;
+}
+
+(** {2 Parsing} *)
+
+val retrieve_log : unit -> raw_event list
+(** [retrieve_log ()] returns the current list of collected [raw_event].*)
+
+module EventParser : sig
```
Comment (denismerigoux) [inline compiler/runtime.mli:182]: Why did you put this inside its own module?

File: compiler/scopelang/scope_to_dcalc.ml [hunk 4]
```diff
@@ -276,11 +276,28 @@ let rec translate_expr (ctx : ctx) (e : Ast.expr Pos.marked) :
         | _ -> e1_func
       in
       let new_args = List.map (translate_expr ctx) args in
+      let input_typ, output_typ =
```
Comment (denismerigoux) [inline compiler/scopelang/scope_to_dcalc.ml:279]: Why did you need the input and output typ here ? Your solution will work for now but it won't be robust when we add functions defined outside scope variables (for instance local let bindings without type information #203 or even anonymous functions). Plus #272 should be merged soon, giving you a way to retrieve the type information in `dcalc` for any sub-expression.

**Review by denismerigoux — COMMENTED** [review]
Looks good for now :)

File: compiler/runtime.ml (hunk 1 above)
Comment (denismerigoux) [reply compiler/runtime.ml:185]: Actually there aren't two kinds. The situation is clearer when you say "When computing variable definitions, there may be intermediate function calls part of the log".

File: compiler/runtime.mli (hunk 3 above)
Comment (denismerigoux) [reply compiler/runtime.mli:182]: That's a fine reason :)

File: compiler/scopelang/scope_to_dcalc.ml (hunk 4 above)
Comment (EmileRolley) [reply compiler/scopelang/scope_to_dcalc.ml:279]: Because without them the input and output variables of functions are unembeddable.

For example, with this workaround the generated OCaml code looks like:
```ocaml
(log_end_call
  ["AllocationsFamiliales"; "droit_ouvert_majoration"; "output"]
  (embed_bool) ((log_begin_call ["AllocationsFamiliales"; "droit_ouvert_majoration"]
    droit_ouvert_majoration_)
  ((log_variable_definition ["AllocationsFamiliales"; "droit_ouvert_majoration"; "input"]
    (embed_enfant) param_))))))
```

instead of:
```ocaml
(log_end_call
  ["AllocationsFamiliales"; "droit_ouvert_majoration"; "output"]
  (unembeddable) ((log_begin_call ["AllocationsFamiliales"; "droit_ouvert_majoration"]
    droit_ouvert_majoration_)
  ((log_variable_definition ["AllocationsFamiliales"; "droit_ouvert_majoration"; "input"]
    (unembeddable) param_))))))
```

File: compiler/runtime.ml (hunk 1 above)
Comment (EmileRolley) [reply compiler/runtime.ml:185]: Okay, I will modify this :ok_hand:

File: compiler/scopelang/scope_to_dcalc.ml (hunk 4 above)
Comment (EmileRolley) [reply compiler/scopelang/scope_to_dcalc.ml:279]: Ok, and maybe refering #272 to be refactored when it's merged?

**Review by denismerigoux — APPROVED** [review]
Looks all good to me, thanks!
